File: prod.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Manejo de Productos"""
import cgitb; cgitb.enable()
import cgi
import funciones
import datos
import pagina
import htm
def listado():
    """Listado de productos"""
    pag = pagina.Pagina("Listado de producotos", 4)
    print(htm.button("Nuevo","prod.py?accion=nuevo"))
    print(htm.button('Volver','geined.py?accion=administracion'))
    # cargar datos
    productos = datos.Tabla("productos")
    productos.orden = "producto"
    productos.filtrar()
    htm.encabezado_tabla(["Nº", "Producto", "Rubro", "Precio", "Acciones"])
    i = 0
    for row in productos.resultado:
        htm.fila_alterna(i)
        print(htm.td(row['id']))
        print(htm.td(row['producto']))
        print(htm.td(row['rubro']))
        print(htm.td(funciones.moneda(row['precio'])))
        print('<td>')
        htm.boton_detalles('prod_ver.php?cuenta_id=' + str(row['id']))
        htm.boton_editar('prod.py?id=' + str(row['id']) + '&accion=editar')
        htm.boton_eliminar('prod.py?accion=confirmar&id=' + str(row['id']))
        print('</td></tr>')
    htm.fin_tabla()
    print(htm.button("Volver", 'geined.py?accion=administracion'))
    pag.fin()
def nuevo():
    """Nuevo producto"""
    pag = pagina.Pagina("Nuevo producto", 4)
    cuentas = datos.Tabla("cuentas")
    cuentas.orden = "rubro"
    cuentas.filtro = "rubro like 4%"
    cuentas.filtrar()
    # voy por aca
    htm.encabezado_tabla(["Campo","Valor"])
    print('<form action="prod.php?accion=agregar" name="f" method="post">')
    htm.input_texto('Producto:','producto','')
    print(htm.tr(htm.td("Rubro:") + htm.td('<input type="text" name="rubro" value="" onChange="codifica()" /> - <input disabled size="80" value="desconocido" name="lblrubro" />')))
    htm.input_numero("Precio:","precio","")
    htm.botones("prod.py?accion=listado")
    htm.fin_formulario()
    htm.fin_tabla()
    print(htm.button('Volver','prod.php?accion=listado'))
    pag.fin()
def agregar(frm):
    productos = datos.Tabla("productos")
    productos.buscar("producto", frm.getvalue("producto"))
    if not productos.encontrado:
        productos.nuevo()
        productos.registro["producto"] = frm.getvalue("producto")
        productos.registro["rubro"] = frm.getvalue("rubro")
        productos.registro["precio"] = frm.getvalue("precio")
        productos.insertar()
        listado()
    else:
        htm.duplicado('prod.py?accion=listado')
def editar(frm):
    """Editar producto"""
    pag = pagina.Pagina("Edición de producto", 4)
    cuentas = datos.Tabla("cuentas")
    cuentas.orden = "rubro"
    cuentas.filtrar()
    productos = datos.Tabla("productos")
    productos.ir_a(frm.getvalue("id"))
    htm.encabezado_tabla(["Campo","Valor"])
    print('<form action="prod.php?accion=actualizar" name="f" method="post">')
    print(htm.hidden("id", frm.getvalue("id")))
    htm.input_texto('Producto:','producto', productos.registro['producto'])
    print(htm.tr(htm.td('Rubro:') + htm.td('<td><input type="text" name="rubro" value="' + productos.registro['rubro'] +  '" onChange="codifica()" /> - <input disabled size="80" value="desconocido" name="lblrubro" />')))
    htm.input_numero("Precio:","precio", productos.registro['precio'])
    htm.botones("prod.py?accion=listado")
    htm.fin_formulario()
    htm.fin_tabla()
    print('<script language="javascript">codifica();</script>')
    print(htm.button('Volver','prod.php?accion=listado'))
    pag.fin()
def actualizar(frm):
    """Actualizar el producto"""
    productos = datos.Tabla("productos")
    productos.ir_a(frm.getvalue("id"))
    productos.registro["rubro"] = frm.getvalue("rubro")
    productos.registro["producto"] = frm.getvalue("producto")
    productos.registro["precio"] = frm.getvalue("precio")
    productos.actualizar()
    listado()
if __name__ == "__main__":
    form = cgi.FieldStorage()
    accion = form.getvalue("accion", "listado")
    if accion == "listado":
        listado()
    elif accion == "editar":
        editar(form)
    elif accion == "nuevo":
        nuevo()
    elif accion == "agregar":
        agregar(form)
    elif accion == "eliminar":
        productos = datos.Tabla("productos")
        productos.borrar(form.getvalue("id"))
        listado()

File: prod.rb
#!/usr/bin/env ruby
require 'cgi_exception'
require 'funciones'
require 'datos'
require 'htm'
def listado
  pag = Pagina.new('Listado de productos', 5)
	boton("Nuevo", "prod.rb?accion=nuevo")
	boton('Volver', 'geined.py?accion=administracion')
	# cargar datos
  productos = Producto.find(:all, :order=>:producto)
	encabezado_tabla(["Nº", "Producto", "Rubro", "Precio", "Acciones"])
  i = 0
	productos.each do |fila|
    fila_alterna(i)
		celda(fila.id)
		celda(fila.producto)
		celda(fila.rubro)
		celda(moneda(fila.precio))
		print '<td>'
		boton("Detalle", "prod_ver.php?cuenta_id=#{fila.id}")
		boton("Editar", "prod.rb?id=#{fila.id}&accion=editar")
		boton("Borrar", "prod.rb?accion=confirmar&id=#{fila.id}")
		puts '</td></tr>'
    i = i + 1
  end
	fin_tabla
	boton("Volver",'geined.py?accion=administracion')
	pag.fin
end
def nuevo
  pag = Pagina.new('Nuevo producto', 4)
  cuentas_combo = Cuenta.find_by_sql('SELECT id,CONCAT(rubro,"-",nombre) as selec FROM cuentas ORDER BY rubro')
  form_edicion('Nuevo producto', 'prod.rb')
  puts pag.cgi.hidden('accion', 'agregar')
	input_texto('Producto:','producto','')
  input_combo()
	puts '<tr><td>Rubro:</td><td><input type="text" name="rubro" value="" onChange="codifica()" /> - <input disabled size="80" value="desconocido" name="lblrubro" /></td></tr>'
	input_numero("Precio:","precio","")
	botones
	fin_formulario
	fin_tabla
	boton('Volver','prod.rb?accion=listado')
end
def agregar(frm)
  begin
    productos = Producto.find(:all, :conditions=>{:producto=>frm['producto']})
    duplicado('prod.rb?accion=listado')
  rescue ActiveRecord::RecordNotFound
    Producto.create(
      :producto => frm['producto'],
      :rubro => frm['rubro'],
      :precio => frm['precio'])
    listado
  end
end
def editar(frm)
	# Entrada
  pag = Pagina.new('Edición de producto', 4)
  cuentas = Cuenta.find(:all, :order=>:rubro)
  producto = Producto.find(frm['id'])
  form_edicion("Editar producto", 'prod.rb')
  puts pag.cgi.hidden('accion', 'actualizar')
  puts pag.cgi.hidden('id', frm['id'])
  input_texto('Producto:', producto.producto)
  input_combo('Rubro:', 'cuenta_id', cuentas, ['id', 'rubro', 'nombre'], frm['cuenta_id'])
	input_numero("Precio:", "precio", frm['precio'])
	botones('prod.rb?accion=listado')
	form_edicion_fin
	pag.fin
end
def actualizar(frm)
  Producto.update(frm['id'],{
    :producto => frm['producto'],
    :cuenta_id => frm['cuenta_id'],
    :precio => frm['precio']
  })
  listado
end
if __FILE__ == $0
  form = CGI.new('html4')
  accion = 'listado'
  if form.has_key?('accion')
    accion = form['accion']
  end
  case accion
  when 'listado'
    listado
  when 'editar'
    editar(form)
  when 'actualizar'
    actualizar(form)
  when 'nuevo'
    nuevo
  when 'agregar'
    agregar(form)
  when 'eliminar'
    Producto.delete(form['id'])
    listado
  end
end